src/StableHashTraits.jl

```diff
@@ -223,113 +224,202 @@ You should return one of the following values.
+- `Tuple`, `Pair`: `Use(qualified_name, UseIterate())`
+- `Type`: `UseQualifiedType`
+- `AbstractArray`: `Use("Base.AbstractArray", UseSize(UseIterate()))`
+- `AbstractRange`: `Use(qualified_name, UseStruct())`
+- `AbstractSet`: `Use(qualified_name, UseTransform(sort! ∘ collect))`
+- `AbstractDict`: `Use(qualified_name, UseStruct(keys => getindex, :ByName))`
+
+There are two built-in contexts that can be used to modify these default fallbacks:
+[`TablesEq`](@ref) and [`ViewsEq`](@ref). `TablesEq` makes any table with equivalent content
+have the same hash, and `ViewsEq` makes any array or string with the same sequence of values
+and the same size have an equal hash. You can pass one or more of these as the second
+argument to `stable_hash`, e.g. `stable_hash(x, ViewsEq())` or `stable_hash(x,
+ViewsEq(TablesEq()))`.
+
+## Customizing hash computations with contexts
+
+You can customize how hashes are computed within a given scope using a context object. This
+is also a very useful way to avoid type piracy. The context can be any object you'd like and
+is passed as the second argument to `stable_hash`. By default it is equal to
+`HashVersion{1}()` and this is the context for which the default fallbacks listed above are
+defined.
+
+This context is then passed to both `hash_method` and `StableHashTraits.write` (the latter
+is the method called for `UseWrite`, and which falls back to `Base.write`). Because of the
+way the default context (`HashVersion{1}`) is defined, you normally don't have to include
+this context as an argument when you define a method of `hash_context` or `write` because
+there are appropriate fallback methods.
+
+When you define a hash context it should normally accept a parent context that serves as a
+fallback, and return it in an implementation of the method
+`StableHashTratis.parent_context`. For example, here is how we could write a context that
+treats all named tuples with the same keys as equivalent. 
 
+```julia
+struct NamedTuplesEq{T}
+    parent::T
+end
+StableHashTraits.parent_context(x::NamedTuplesEq) = x.parent
+function StableHashTraits.hash_method(::NamedTuple, ::NamedTuplesEq) 
+    return Use(qualified_name, UseStruct(:ByName))
+end
+c = NamedTuplesEq(HashVersion{1}())
+stable_hash((; a=1:2, b=1:2), c) == stable_hash((; b=1:2, a=1:2), c) # true
 ```
+
+If we did not define a method of `parent_context`, our context would need to implement a
+`hash_method` that covered the types `AbstractRange`, `Int64`, `Symbol` and `Pair` for the
+call to `stable_hash` above to succeede.
+
+### Customizing hashes within an object
+
+Contexts can be changed not only when you call `stable_hash` but also when you hash the
+contents of a particular object. This lets you change how hashing occurs within the object.
+See the docstring of [`UseAndReplaceContext`](@ref) for details. 
+"""
+hash_method(x, context) = hash_method(x, parent_context(context))
+hash_method(x, ::Nothing) = hash_method(x)
+hash_method(::Any) = nothing
```

Comment:
remove 1-argument `hash_method` methods

Reply:
Thinking on this more I realized the reall problem is with method ambiguities. I've update the comments to describe this more precisely for anyone wondering why this is the way it works.